File: work/SmartLearning-Backend/src/main/java/Final/year/project/SmartLearning/Enrollments/EnrollmentService.java
package Final.year.project.SmartLearning.Enrollments;

import Final.year.project.SmartLearning.dto.EnrollRequest;
import Final.year.project.SmartLearning.dto.EnrollmentResponse;

import java.util.List;
import java.util.UUID;

public interface EnrollmentService {

    EnrollmentResponse enrollStudent(
            EnrollRequest request);

    void unenrollStudent(UUID courseId);

    boolean isEnrolled(UUID courseId);

    List<EnrollmentResponse> getMyCourses();

    List<EnrollmentResponse> getEnrollmentsByStudent(
            UUID studentId);

    List<EnrollmentResponse> getEnrollmentsByCourse(
            UUID courseId);

    void removeEnrollment(UUID enrollmentId);

    /**
     * CSV export of the course roster, including each student's score so far
     * (sum of graded submissions in this course). Lecturer must own the
     * course; admin/master may export any course.
     */
    String exportRosterCsv(UUID courseId);
}

File: work/Smartlearning-Frontend/src/app/core/services/enrollment.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { EnrollRequest, EnrollmentResponse } from '../models/models';

@Injectable({ providedIn: 'root' })
export class EnrollmentService {
  private readonly apiUrl = `${environment.apiUrl}/enrollments`;

  constructor(private http: HttpClient) {}

  enroll(request: EnrollRequest): Observable<EnrollmentResponse> {
    return this.http.post<EnrollmentResponse>(`${this.apiUrl}/enroll`, request);
  }

  unenroll(courseId: string): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/unenroll/${courseId}`);
  }

  checkEnrollment(courseId: string): Observable<boolean> {
    return this.http.get<boolean>(`${this.apiUrl}/check/${courseId}`);
  }

  getMyCourses(): Observable<EnrollmentResponse[]> {
    return this.http.get<EnrollmentResponse[]>(`${this.apiUrl}/my-courses`);
  }

  getEnrollmentsByStudent(studentId: string): Observable<EnrollmentResponse[]> {
    return this.http.get<EnrollmentResponse[]>(`${this.apiUrl}/student/${studentId}`);
  }

  getEnrollmentsByCourse(courseId: string): Observable<EnrollmentResponse[]> {
    return this.http.get<EnrollmentResponse[]>(`${this.apiUrl}/course/${courseId}`);
  }

  removeEnrollment(id: string): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  /** Downloads a CSV roster for a course - every enrolled student with their total score. */
  exportRosterCsv(courseId: string): Observable<Blob> {
    return this.http.get(`${this.apiUrl}/course/${courseId}/export`, { responseType: 'blob' });
  }
}
